pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java

```diff
@@ -355,25 +355,8 @@ private int getRestrictedMaxEntriesForConsumer(Consumer consumer, List<Entry> en
             return maxMessages;
         }
 
-        // If the read type is Replay, we should avoid send messages that hold by other consumer to the new consumers,
-        // For example, we have 10 messages [0,1,2,3,4,5,6,7,8,9]
-        // If the consumer0 get message 0 and 1, and does not acked message 0, then consumer1 joined,
-        // when consumer1 get message 2,3, the broker will not dispatch messages to consumer1
-        // because of the mark delete position did not move forward.
-        // So message 2,3 will stored in the redeliver tracker.
-        // Now, consumer2 joined, it will read new messages from the cursor,
-        // so the recentJoinedPosition is 4 for consumer2
-        // Because of there are messages need to redeliver, so the broker will read the redelivery message first [2,3]
-        // message [2,3] is lower than the recentJoinedPosition 4,
-        // so the message [2,3] will dispatched to the consumer2
-        // But the message [2,3] should not dispatch to consumer2.
-
         if (readType == ReadType.Replay) {
```

Comment:
> However, the key-shared subscription is incorrectly handling redelivered messages by reading redelivered messages, discarding them, and not dispatching any single messages to the consumer by incorrectly changing the semantics of the consumer delivery order. Broker doesn't dispatch redelivery message if that message-id is smaller than the consumer's assigned offset-message-id when joined.

These codes guarantee the ordering of messages during the scenario below:

| No. | Consumer 1 | Consumer 2 | Consumer 3 | Consumer 4 |
| --- | --- | --- | --- | --- |
| stat | handling `k1,k2`, `recent-join: null`  | handling `k3,k4`, `recent-join: null` |
| 1 | received `M1(k1), M2(k2)`| received `1000` messages (`M3(k3)...M1002(k3)`) |
| 2 | | | added | 
| description | | | assigned `k2` which from Consumer 1 |
| stat | handling `k1` | handling `k3,k4`| handling `k2`, `recent-join: M1002` |
| 3 | | closed | 
| description | assigned `k3` which from Consumer 2 | |  assigned `k4` which from Consumer 2 |
| stat | handling `k1,k3`, `recent-join: null` | | handling `k2, k4`, `recent-join: M1002` |
| 4 | received `M3(k3)...M1000(k3)`|
| 5 | | | | | added |
| description |  | |  | assigned `k3` which from Consumer 1 |
| state |handling `k1`, `recent-join: null` | | handling `k2, k4`, `recent-join: M1002` | handling `k3`, `recent-join: M1002` |
| 6 | | | | received `M1001(k3)...M1002(k3)`|

I think we should solve the issue above first, then try to improve here.

Related to https://github.com/apache/pulsar/pull/20776, please take a look

Reply:
@poorbarcode can you share the URL where we have defined the contract of key-shared sub.

Reply:
> @poorbarcode can you share the URL where we have defined the contract of the key-shared sub?

I do not know what URL you wanted, is the doc of [Key_Share Subscription Doc](https://pulsar.apache.org/docs/3.1.x/client-libraries-consumers/#key_shared)?

Reply:
@poorbarcode I mean do we have any document where we have shared what kind of ordering guarantee we provide to users? because as I said in the issue, once one consumer is closed, broker can redeliver unack messages of that consumers without considering the ordering instead blocking forever.
So, I just want to see if we have any doc where we have defined what user can expect in terms ordering for key-shared sub. I had checked earlier URL which you shared but that doesn't talk about the ordering or redelivery ordering.
In this PR with latest commit, It maintains the ordering guarantee but it also handles redelivery of unack messages of closed consumer without blocking dispatcher forever.

So, if we have any contract defined then we can check if this PR violates the user contract for the key-shared subscription because right now, key-shared sub is not usable and it is wasting lot of broker/bookie resources.

Reply:
[Preserving order of processing](https://pulsar.apache.org/docs/3.1.x/concepts-messaging/#preserving-order-of-processing)

Reply:
thanks @Technoboy- for sharing the link and I was searching for this documentation where it talks about message ordering guarantee for key-shared.

>The broker will start delivering messages to the new consumer only when all messages up to the read position have been acknowledged. This will guarantee that a certain key is processed by a single consumer at any given time. The trade-off is that if one of the existing consumers is stuck and no time-out was defined (acknowledging for you), the new consumer won't receive any messages until the stuck consumer resumes or gets disconnected.

Right now, dispatching gets stuck when consumer gets disconnected and this PR has the test to reproduce it.
and this PR exactly fixes that issue to unblock dispatching if consumer disconnects and redeliver that consumer's unack messages.

So, this PR should fix that fundamental issue to unblock stuck consumers when they should not be stuck.

Reply:
> So, this PR should fix that fundamental issue to unblock stuck consumers when they should not be stuck.

Great points @rdhabalia . We can continue to resolve this issue as part of PIP-379, #23309 is the PIP document.

Reply:
@lhotari you have updated a contract similar to this PR and tried the same behavior in #23309 . then why this PR was blocked for 5 months and then you closed it?
I really don't have words to mention what's going on. does it really look good to people who blocked the PR and came up with a similar approach and did not let other people's work move forward? You also know that the same things keep happening again and again in other PRs as well and you have also witnessed this kind of thing in Pulsar very recently. 
@lhotari I don't want to target anyone here but want to ask a simple question: does it look good to do such kind of actions? does it make any difference to their lives by doing it? Because I really don't understand what's going on in this Project recently.